src/openboardview/FileFormats/BVR3File.cpp

```diff
@@ -0,0 +1,207 @@
+#include "BVR3File.h"
+
+#include "utils.h"
+#include <cmath>
+#include <cctype>
+#include <clocale>
+#include <cstdint>
+#include <cstring>
+#include <list>
+
+int manhattan_distance(const BRDPoint &p1, const BRDPoint &p2) {
+	return abs(p1.x - p2.x) + abs(p1.y - p2.y);
+}
+
+bool BVR3File::verifyFormat(std::vector<char> &buf) {
+	return find_str_in_buf("BVRAW_FORMAT_3", buf);
+}
+
+BVR3File::BVR3File(std::vector<char> &buf) {
+	auto buffer_size = buf.size();
+
+	char *saved_locale = setlocale(LC_NUMERIC, "C"); // Use '.' as delimiter for strtod
+
+	ENSURE_OR_FAIL(buffer_size > 4, error_msg, return);
+	size_t file_buf_size = 3 * (1 + buffer_size);
+	file_buf             = (char *)calloc(1, file_buf_size);
+	ENSURE_OR_FAIL(file_buf != nullptr, error_msg, return);
+
+	std::copy(buf.begin(), buf.end(), file_buf);
+	file_buf[buffer_size] = 0;
+	// This is for fixing degenerate utf8
+	char *arena     = &file_buf[buffer_size + 1];
+	char *arena_end = file_buf + file_buf_size - 1;
+	*arena_end      = 0;
+
+	BRDPart blank_part;
+	BRDPin blank_pin;
+	BRDPart part;
+	BRDPin pin;
+	std::list<std::pair<BRDPoint, BRDPoint>> outline_segments;
+
+	std::vector<char *> lines;
+	stringfile(file_buf, lines);
+
+	for (char *line : lines) {
+		while (isspace((uint8_t)*line)) line++;
+		if (!line[0]) continue;
+
+		char *p = line;
+		char *s;
+
+		if (!strncmp(line, "PART_NAME ", 10)) {
+			p += 10;
+			part.name = READ_STR();
+		} else if (!strncmp(line, "PART_SIDE ", 10)) {
+			p += 10;
+			char *side = READ_STR();
+			if (!strcmp(side, "T"))
+				part.mounting_side = BRDPartMountingSide::Top;
+			else
+				part.mounting_side = BRDPartMountingSide::Bottom;
+		} else if (!strncmp(line, "PART_ORIGIN ", 12)) {
+			p += 12;
+			double origin_x = READ_DOUBLE();
+			part.p1.x = trunc(origin_x);
```

Comment:
Note that `BRDPart::p1` and `BRDPart::p2` are only used to draw a rectangle for parts that have no pins, and for that purpose, both points have to be defined.
 I don't know what `PART_ORIGIN` is supposed to mean since it's almost always 0 except in some files with "orphan" probe points (or vias) where it is set to the position of the pin.
For now maybe better to leave it out.

Reply:
Have asked what the possible values are for this in #224. I agree though that it's probably better to leave it out unless the meaning highlights something we don't know.

Reply:
This is apparently an origin point used when relative points for pins are defined. In any case there is no support for this currently so the origin point is also not useful at present. I have removed this code and added a comment to explain.